datafusion/physical-expr/src/aggregate/bit_and_or_xor.rs

```diff
@@ -0,0 +1,1103 @@
+        }
+    }
+}
+
+// Bit and/Bit or/Bit xor aggregation can take Dictionary encode input but always produces unpacked
+// (aka non Dictionary) output. We need to adjust the output data type to reflect this.
+// The reason bit and/bit or/bit xor aggregate produces unpacked output because there is only one
+// bit and/bit or/bit xor value per group; there is no needs to keep them Dictionary encode
+fn bit_and_or_xor_aggregate_data_type(input_type: DataType) -> DataType {
+    if let DataType::Dictionary(_, value_type) = input_type {
+        *value_type
+    } else {
+        input_type
+    }
+}
+
+// returns the new value after bit_and/bit_or/bit_xor with the new values, taking nullability into account
+macro_rules! typed_bit_and_or_xor_batch {
+    ($VALUES:expr, $ARRAYTYPE:ident, $SCALAR:ident, $OP:ident) => {{
+        let array = downcast_value!($VALUES, $ARRAYTYPE);
+        let delta = $OP(array);
+        Ok(ScalarValue::$SCALAR(delta))
+    }};
+}
+
+// bit_and/bit_or/bit_xor the array and returns a ScalarValue of its corresponding type.
+macro_rules! bit_and_or_xor_batch {
+    ($VALUES:expr, $OP:ident) => {{
+        match $VALUES.data_type() {
+            DataType::Int64 => {
+                typed_bit_and_or_xor_batch!($VALUES, Int64Array, Int64, $OP)
+            }
+            DataType::Int32 => {
+                typed_bit_and_or_xor_batch!($VALUES, Int32Array, Int32, $OP)
+            }
+            DataType::Int16 => {
+                typed_bit_and_or_xor_batch!($VALUES, Int16Array, Int16, $OP)
+            }
+            DataType::Int8 => typed_bit_and_or_xor_batch!($VALUES, Int8Array, Int8, $OP),
+            DataType::UInt64 => {
+                typed_bit_and_or_xor_batch!($VALUES, UInt64Array, UInt64, $OP)
+            }
+            DataType::UInt32 => {
+                typed_bit_and_or_xor_batch!($VALUES, UInt32Array, UInt32, $OP)
+            }
+            DataType::UInt16 => {
+                typed_bit_and_or_xor_batch!($VALUES, UInt16Array, UInt16, $OP)
+            }
+            DataType::UInt8 => {
+                typed_bit_and_or_xor_batch!($VALUES, UInt8Array, UInt8, $OP)
+            }
+            e => {
+                return Err(DataFusionError::Internal(format!(
+                    "Bit and/Bit or/Bit xor is not expected to receive the type {e:?}"
+                )));
+            }
+        }
+    }};
+}
+
+/// dynamically-typed bit_and(array) -> ScalarValue
+fn bit_and_batch(values: &ArrayRef) -> Result<ScalarValue> {
+    bit_and_or_xor_batch!(values, bit_and)
+}
+
+/// dynamically-typed bit_or(array) -> ScalarValue
+fn bit_or_batch(values: &ArrayRef) -> Result<ScalarValue> {
+    bit_and_or_xor_batch!(values, bit_or)
+}
+
+/// dynamically-typed bit_xor(array) -> ScalarValue
+fn bit_xor_batch(values: &ArrayRef) -> Result<ScalarValue> {
+    bit_and_or_xor_batch!(values, bit_xor)
+}
+
+// bit_and/bit_or/bit_xor of two scalar values.
+macro_rules! typed_bit_and_or_xor {
```

Comment:
I recommend we make this a method on `ScalarValue` so that it is more discoverable